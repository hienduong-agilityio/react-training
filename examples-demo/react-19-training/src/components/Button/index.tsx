// Libraries
import { memo } from 'react';
import { twMerge } from 'tailwind-merge';
import classNames from 'classnames';

// Types
import type { ButtonHTMLAttributes, JSX, ReactNode } from 'react';
import { Spinner } from '../../icons';

export enum BUTTON_VARIANTS {
  'SOLID' = 'solid',
  'SHADOW' = 'shadow',
  'LIGHT' = 'light'
}

export enum BUTTON_COLORS {
  'DEFAULT' = 'default',
  'PRIMARY' = 'primary',
  'SECONDARY' = 'secondary',
  'SUCCESS' = 'success',
  'WARNING' = 'warning',
  'DANGER' = 'danger'
}

// Icons

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode;
  startIcon?: ReactNode;
  endIcon?: ReactNode;
  variant?: BUTTON_VARIANTS;
  color?: BUTTON_COLORS;
  customClass?: string;
  isLoading?: boolean;
  loadingText?: ReactNode;
}

const baseClass = 'flex items-center justify-center gap-2 rounded p-2 disabled:cursor-not-allowed disabled:opacity-50';

const variantClasses: Record<BUTTON_VARIANTS, string> = {
  [BUTTON_VARIANTS.SOLID]: 'bg-opacity-100 shadow-md',
  [BUTTON_VARIANTS.SHADOW]: 'shadow-lg bg-white hover:shadow-xl',
  [BUTTON_VARIANTS.LIGHT]: 'bg-gray-100 bg-opacity-50 hover:bg-opacity-70'
};

const colorClasses: Record<BUTTON_COLORS, string> = {
  [BUTTON_COLORS.DEFAULT]: 'bg-gray-200 text-black',
  [BUTTON_COLORS.PRIMARY]: 'bg-primary-200 text-white disabled:hover:bg-primary-200 hover:bg-primary-400',
  [BUTTON_COLORS.SECONDARY]: 'bg-gray-400 text-white',
  [BUTTON_COLORS.SUCCESS]: 'bg-green-600 text-white',
  [BUTTON_COLORS.WARNING]: 'bg-amber-400 text-black',
  [BUTTON_COLORS.DANGER]: 'bg-rose-500 text-white'
};

/**
 * Button component
 *
 * @returns {JSX.Element} - The rendered button element.
 */
const Button = ({
  children,
  startIcon,
  endIcon = '',
  type = 'button',
  variant = BUTTON_VARIANTS.SOLID,
  color = BUTTON_COLORS.DEFAULT,
  customClass = '',
  isLoading = false,
  loadingText = '',
  ...rest
}: IButtonProps): JSX.Element => {
  const combinedClasses = twMerge(classNames(baseClass, variantClasses[variant], colorClasses[color]), customClass);

  return (
    <button type={type} className={combinedClasses} disabled={isLoading || rest.disabled} {...rest}>
      {isLoading ? (
        <>
          <Spinner size={20} color='currentColor' />
          {loadingText || children}
        </>
      ) : (
        <>
          {startIcon && <span>{startIcon}</span>}
          {children}
          {endIcon && <span>{endIcon}</span>}
        </>
      )}
    </button>
  );
};

export default memo(Button);
