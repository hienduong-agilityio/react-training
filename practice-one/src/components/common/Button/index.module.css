.btn {
  background-color: var(--primary-color);
  border-radius: 0.25rem;
}

.primary {
  background-color: var(--primary-color);
  color: var(--white-color);
}

.secondary {
  background-color: var(--active-gray-color);
}

.outline {
  border: none;
}

.icon {
  background-color: var(--icon-color);
}

