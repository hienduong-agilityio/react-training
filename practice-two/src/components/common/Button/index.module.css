.button {
  display: inline-flex;
  padding: 0rem 0.5rem;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  flex-shrink: 0;
  border: none;
  border-radius: 0.25rem;
  cursor: pointer;
}

.text {
  background-color: var(--white-color);
}

.text:hover {
  background-color: var(--shadow-color);
}

.outline {
  border: 1px solid var(--primary-color);
}

.outline:hover {
  box-shadow: 0px 1.6px 3.6px 0px rgba(0, 0, 0, 0.11);
}

.filled {
  background-color: var(--shadow-color);
}

.filledTonal {
  background-color: var(--shadow-color);
  border: 1px solid var(--primary-color);
}

.primary {
  background-color: var(--primary-color);
  color: var(--white-color);
}

.small {
  height: 1.5rem;
}

.medium {
  height: 2.5rem;
}

.large {
  height: 3.5rem;
}

.light {
  background-color: var(--white-color);
}

.icon {
  background-color: var(--icon-color);
}
