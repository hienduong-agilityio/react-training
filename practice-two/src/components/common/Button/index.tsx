// Library
import { ButtonHTMLAttributes } from 'react';
import classNames from 'classnames';

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'text' | 'outline' | 'filled' | 'filledTonal';
  size?: 'small' | 'medium' | 'large';
  btnType?: 'button' | 'submit' | 'reset';
  color?: 'primary' | 'secondary' | 'default' | 'danger';
  customClasses?: string;
}

export enum BUTTON_VARIANT {
  TEXT = 'text',
  OUTLINE = 'outline',
  FILLED = 'filled',
  FILLED_TONAL = 'filledTonal'
}

const buttonDefault: string =
  'rounded-2xl justify-center font-semibold flex gap-2 items-center hover:opacity-80 active:shadow-buttonActive cursor-pointer disabled:opacity-60';

const variantClasses: { [key in BUTTON_VARIANT]: string } = {
  text: 'hover:bg-gray-200',
  outline: 'bg-white border',
  filled: 'text-white',
  filledTonal: 'text-white shadow-mg hover:shadow-lg'
};

const sizeClasses: { [key in 'small' | 'medium' | 'large']: string } = {
  small: 'px-3 py-1 text-sm',
  medium: 'px-4 py-2 text-base',
  large: 'px-6 py-3 text-lg'
};

const colorClasses: { [key in 'default' | 'primary' | 'secondary' | 'danger']: string } = {
  default: 'text-primary border-primary',
  primary: 'bg-primary border-primary',
  secondary: 'bg-secondary border-secondary',
  danger: 'bg-danger border-danger'
};

/**
 * Button component
 * @param variant Variant of the button. Can be one of 'text', 'outline', 'filled', or 'filledTonal'.
 * @param size Size of the button. Can be one of 'small', 'medium', or 'large'.
 * @param color Color of the button. Can be one of 'primary', 'secondary', 'default'.
 * @param customClasses Custom variant class for the button.
 *
 * @returns {JSX.Element} - Button element.
 */

const Button = ({
  children,
  variant = 'text',
  size = 'medium',
  color = 'default',
  customClasses = '',
  ...restProps
}: IButtonProps): JSX.Element => {
  const buttonClasses: string = classNames([
    buttonDefault,
    variantClasses[variant],
    sizeClasses[size],
    colorClasses[color],
    customClasses
  ]);

  return (
    <button className={buttonClasses} {...restProps}>
      {children}
    </button>
  );
};

export default Button;
