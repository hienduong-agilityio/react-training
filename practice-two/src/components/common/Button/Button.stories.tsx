import type { Meta, StoryObj } from '@storybook/react';
import Button, { BUTTON_VARIANT } from './index';

const meta = {
  title: 'Components/common/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs']
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    children: 'Button primary',
    variant: 'text',
    customClasses: 'bg-blue-400',
    disabled: true,
    onClick: () => {
      alert('Button component');
    }
  }
};

export const Outline: Story = {
  args: {
    children: 'Button secondary',
    variant: BUTTON_VARIANT.OUTLINE,
    onClick: () => {
      alert('Button component');
    }
  }
};

export const Filled: Story = {
  args: {
    children: 'Button filled',
    variant: BUTTON_VARIANT.FILLED,
    onClick: () => {
      alert('Button component');
    }
  }
};
