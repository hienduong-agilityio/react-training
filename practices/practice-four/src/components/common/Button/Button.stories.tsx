// Types
import type { Meta, StoryObj } from '@storybook/react';

// Enums
import { BUTTON_VARIANTS, BUTTON_SIZES, BUTTON_COLORS } from '@/enums';

// Component
import Button from './index';

const meta = {
  title: 'Components/common/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs'],
  argTypes: {
    variant: {
      description: 'Variant of the button. Can be one of "text", "outline", "contained".',
      control: { type: 'select', options: Object.values(BUTTON_VARIANTS) }
    },
    size: {
      description: 'Size of the button. Can be one of "small", "medium", or "large".',
      control: { type: 'select', options: Object.values(BUTTON_SIZES) }
    },
    color: {
      description: 'Color of the button. Can be one of "primary", "default".',
      control: { type: 'select', options: Object.values(BUTTON_COLORS) }
    },
    customClasses: {
      description: 'Custom CSS classes to be applied to the button.',
      control: { type: 'text' }
    },
    onClick: {
      description: 'Function to handle button click event.',
      action: 'clicked'
    },
    children: {
      description: 'Content inside the button.',
      control: { type: 'text' }
    }
  }
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    customClasses: 'm-5'
  },

  render: (args) => (
    <>
      <Button {...args}>Default</Button>
      <Button disabled {...args}>
        Default
      </Button>
    </>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Default button with optional disabled state.'
      }
    }
  }
};

export const ContainedVariant: Story = {
  args: {
    children: 'Button primary',
    type: 'button',
    variant: BUTTON_VARIANTS.CONTAINED,
    customClasses: 'm-5',

    onClick: () => {
      alert('Button component');
    }
  },

  render: (args) => (
    <>
      <Button color={BUTTON_COLORS.PRIMARY} {...args}>
        Primary
      </Button>
      <Button color={BUTTON_COLORS.DANGER} {...args}>
        Danger
      </Button>
    </>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Button with a "contained" variant, showcasing primary and danger colors.'
      }
    }
  }
};

export const OutlineVariant: Story = {
  args: {
    children: 'Button secondary',
    variant: BUTTON_VARIANTS.OUTLINED,
    onClick: () => {
      alert('Button component');
    }
  },
  parameters: {
    docs: {
      description: {
        story: 'Button with an "outlined" variant.'
      }
    }
  }
};

export const TextVariant: Story = {
  args: {
    children: 'Button text',
    variant: BUTTON_VARIANTS.TEXT,
    onClick: () => {
      alert('Button component');
    }
  },
  parameters: {
    docs: {
      description: {
        story: 'Button with a "text" variant.'
      }
    }
  }
};

export const TextWithIcon: Story = {
  args: {
    children: 'Text With Icon',
    variant: BUTTON_VARIANTS.CONTAINED,
    color: BUTTON_COLORS.PRIMARY,
    onClick: () => {
      alert('Button component');
    }
  },
  render: (args) => {
    return (
      <Button {...args}>
        <svg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16' fill='#ffff'>
          <path
            fill-rule='evenodd'
            clip-rule='evenodd'
            d='M7.99999 2C8.4142 2 8.74999 2.33579 8.74999 2.75V7.25H13.25C13.6642 7.25 14 7.58579 14 8C14 8.41422 13.6642 8.75 13.25 8.75H8.74999V13.25C8.74999 13.6642 8.4142 14 7.99999 14C7.58578 14 7.24999 13.6642 7.24999 13.25V8.75H2.75C2.33579 8.75 2 8.41422 2 8C2 7.58579 2.33579 7.25 2.75 7.25H7.24999V2.75C7.24999 2.33579 7.58578 2 7.99999 2Z'
            fill='#FFFFF'
          />
        </svg>
        Button
      </Button>
    );
  },
  parameters: {
    docs: {
      description: {
        story: 'Button with text and an icon, using the "contained" variant and primary color.'
      }
    }
  }
};

export const ButtonWithDifferentSize: Story = {
  args: {
    variant: BUTTON_VARIANTS.TEXT,
    customClasses: 'inline-flex'
  },

  render: (args) => (
    <>
      <Button size={BUTTON_SIZES.SMALL} {...args}>
        Small
      </Button>
      <Button size={BUTTON_SIZES.MEDIUM} {...args}>
        Medium
      </Button>
      <Button size={BUTTON_SIZES.LARGE} {...args}>
        Large
      </Button>
    </>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Button with different sizes: small, medium, and large.'
      }
    }
  }
};
