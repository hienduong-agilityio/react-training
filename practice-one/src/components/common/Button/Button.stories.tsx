import type { Meta, StoryObj } from '@storybook/react';
import Button from './Button';
import { BUTTON_VARIANT } from '../../../types/common';

const meta = {
  title: 'Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    children: 'Button primary',
    variant: BUTTON_VARIANT.PRIMARY,
    className: 'outline',
    onClick: () => {
      alert('Button component');
    },
  },
};

export const Secondary: Story = {
  args: {
    children: 'Button secondary',
    variant: BUTTON_VARIANT.SECONDARY,
    className: '',
    onClick: () => {
      alert('Button component');
    },
  },
};

export const Outline: Story = {
  args: {
    children: 'Button outline',
    variant: BUTTON_VARIANT.OUTLINE,
    className: '',
    onClick: () => {
      alert('Button component');
    },
  },
};
