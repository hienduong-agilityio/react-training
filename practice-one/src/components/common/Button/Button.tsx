import { BUTTON_VARIANT } from '../../../types/common';
import style from './index.module.css';

export interface ButtonProps {
  children: string;
  variant?: BUTTON_VARIANT;
  className: string;
  onClick?: () => void;
}

const Button = ({ children, variant = BUTTON_VARIANT.PRIMARY, className, onClick }: ButtonProps) => {
  return (
    <button type="button" onClick={onClick} className={`${style.button} ${style[variant]} ${style[className]}`}>
      {children}
    </button>
  );
};

export default Button;
