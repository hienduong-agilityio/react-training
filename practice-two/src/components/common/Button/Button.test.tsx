import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from './index';

describe('Button Component', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with custom props', () => {
    const { container } = render(
      <Button
        variant="filled"
        size="large"
        color="primary"
        disabled={false}
        customClasses="extra-class"
        leftIcon={<span>Left Icon</span>}
        rightIcon={<span>Right Icon</span>}
        onClick={() => {}}
      >
        Click me
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls onClick handler when clicked', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick}>Click me</Button>
    );
    fireEvent.click(getByText('Click me'));
    expect(handleClick).toHaveBeenCalled();
  });

  it('disables button when disable prop is true', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button disabled={true} onClick={handleClick}>
        Click me
      </Button>
    );
    const button = getByText('Click me');
    expect(button).toBeDisabled();
    fireEvent.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });
});
