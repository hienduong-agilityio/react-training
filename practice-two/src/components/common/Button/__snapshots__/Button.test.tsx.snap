// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly with custom props 1`] = `
<button
  class="inline-flex rounded-xl font-medium uppercase flex gap-2 w-max items-center hover:opacity-80 active:shadow-buttonActive cursor-pointer disabled:opacity-60 bg-primary text-white rounded-lg px-6 py-3 text-lg bg-primary extra-class"
  type="button"
>
  <span
    class="inline-flex rounded-xl font-medium uppercase flex gap-2 w-max items-center hover:opacity-80 active:shadow-buttonActive cursor-pointer disabled:opacity-60 bg-primary text-white rounded-lg px-6 py-3 text-lg bg-primary extra-class"
  >
    <span>
      Left Icon
    </span>
  </span>
  Click me
  <span
    class="inline-flex rounded-xl font-medium uppercase flex gap-2 w-max items-center hover:opacity-80 active:shadow-buttonActive cursor-pointer disabled:opacity-60 bg-primary text-white rounded-lg px-6 py-3 text-lg bg-primary extra-class"
  >
    <span>
      Right Icon
    </span>
  </span>
</button>
`;

exports[`Button Component renders correctly with default props 1`] = `
<button
  class="inline-flex rounded-xl font-medium uppercase flex gap-2 w-max items-center hover:opacity-80 active:shadow-buttonActive cursor-pointer disabled:opacity-60 bg-primary hover:bg-blue-700 border-blue-600 border hover:border-blue-700 px-4 py-2 text-base"
  type="button"
>
  Click me
</button>
`;
